src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1157,6 +1160,121 @@ static bool isMaybeDefaultValue(TypeWithState valueType)
             }
         }
 
+        internal static bool AreParameterAnnotationsCompatible(
+                RefKind refKind,
+                TypeWithAnnotations overriddenType,
+                FlowAnalysisAnnotations overriddenAnnotations,
+                TypeWithAnnotations overridingType,
+                FlowAnalysisAnnotations overridingAnnotations,
+                bool forRef = false)
+        {
+            // We've already checked types and annotations, let's check nullability attributes as well
+            // Return value is treated as an `out` parameter (or `ref` if it is a `ref` return)
+
+            if (refKind == RefKind.Ref)
+            {
+                // ref variables are invariant
+                return AreParameterAnnotationsCompatible(RefKind.None, overriddenType, overriddenAnnotations, overridingType, overridingAnnotations, forRef: true) &&
```

Comment:
Why do we pass in `RefKind.None` and also `forRef: true`? What distinguishes the meaning of those parameters?

Reply:
On a `RefKind.None` parameter, `[NotNull]` takes a special meaning. We don't want that meaning to apply to a `ref` argument when it is analyzed as `RefKind.None` + `RefKind.Out`.